engineio/src/socket.rs

```diff
@@ -165,6 +179,26 @@ impl Socket {
         Ok(())
     }
 
+    // Returns the time in milliseconds that is left until a new ping must be received.
+    // This is used to detect whether we have been disconnected from the server.
+    // See https://socket.io/docs/v4/how-it-works/#disconnection-detection
+    fn time_to_next_ping(&self) -> u64 {
+        match self.last_ping.lock() {
+            Ok(last_ping) => match Instant::now().checked_duration_since(*last_ping) {
+                Some(since_last_ping) => {
+                    let since_last_ping = (since_last_ping.as_millis() >> 64) as u64;
+                    if since_last_ping > self.max_ping_timeout {
+                        0
+                    } else {
+                        self.max_ping_timeout - since_last_ping
+                    }
+                }
+                None => 0,
+            },
+            Err(_) => 0,
```

Comment:
Maybe a panic would be more appropriate here as the Mutex would be poisoned in this state.

Reply:
Yeah, I'll just make this function return a `Result` and use `*self.last_ping.lock()?` just like everywhere else.

Reply:
But isn't a poisoned mutex a reason to fail completely?

Reply:
Hm, the documentation of the `lock` function says:

> If another user of this mutex panicked while holding the mutex, then this call will return an error once the mutex is acquired.

Therefore, when this call errors, there must be someone else that panicked already, no?

But either way, there is the `InvalidPoisonedLock` (`engineio/src/errors.rs`) error already which is created from `std::sync::PoisonError`, so I think this should be used for now.
Also, because that is the way it's currently done and I'd like to not introduce a different behavior for now :)

Reply:
Yeah, but the mutex can never be acquired again.